rule/utils.go

```diff
@@ -140,3 +141,76 @@ func isCallToExitFunction(pkgName, functionName string) bool {
 func newInternalFailureError(e error) []lint.Failure {
 	return []lint.Failure{lint.NewInternalFailure(e.Error())}
 }
+
+// isASCIIUpper checks if rune is ascii upper case letter
+func isASCIIUpper(r rune) bool {
+	return r >= 'A' && r <= 'Z'
+}
+
+// hasUpperCaseLetter checks if string contains at least one upper case letter
+func hasUpperCaseLetter(s string) bool {
+	for _, r := range s {
+		if isASCIIUpper(r) {
+			return true
+		}
+	}
+	return false
+}
+
+// isUpperCaseConst checks if string is in constant name format like `SOME_CONST`, `SOME_CONST_2`, `X123_3`, `_SOME_PRIVATE_CONST` (#851, #865)
+func isUpperCaseConst(s string) bool {
+	if len(s) == 0 {
+		return false
+	}
+	r := []rune(s)
+	c := r[0]
+	if len(r) == 1 {
+		return isASCIIUpper(c)
+	}
+	switch {
+	case c == '_', isASCIIUpper(c):
+		// Must start with an uppercase letter or underscore
+	default:
+		return false
+	}
+	for i, c := range r {
+		switch {
+		case isUpperOrDigit(c):
+			continue
+		case c == '_':
+			// Underscore must be followed by at least one uppercase letter or digit
+			if i+1 >= len(s) || !isUpperOrDigit(r[i+1]) {
+				return false
+			}
+
+		default:
+			return false
+		}
+	}
+	return true
+}
+
+// isUpperOrDigit checks if rune is upper case letter or digit
+func isUpperOrDigit(c rune) bool {
+	return isASCIIUpper(c) || unicode.IsDigit(c)
+}
```

Comment:
I propose to create `isDigit` function instead of `unicode.IsDigit` because `unicode.IsDigit('৩') == true`.
```suggestion
func isUpperOrDigit(r rune) bool {
	return isUpper(r) || isDigit(c)
}
```

Also, let's use `r` for consistency with `isUpper`.